### PR TITLE
fix(app): reconcile saved story objective ids with the published catalog

### PR DESCRIPTION
**Pull request by DysektAI**
## What

Stacked on #849 (base is `fix/storyline-overlay-contract`; GitHub retargets this to `main` once that merges). Third piece of the overlay v1.93 rollout, alongside [tarkov-data-overlay#377](https://github.com/tarkovtracker-org/tarkov-data-overlay/pull/377).

Overlay v1.93 replaces The Ticket's 44 hand-authored positional objective IDs with real client IDs. Nothing in TarkovTracker prunes or re-keys saved objective marks, so without this change every one of those records:

- stops resolving, and the player sees the objective as untracked,
- stays in the progress blob permanently — sanitizers pass unknown keys through untouched,
- is re-uploaded on every sync and re-saved beside the mark the player makes again under the new ID,
- takes the prestige 4/5 gate with it: the overlay moved that `storyObjectiveStatus` requirement to the new ID, so a player who already recorded the milestone reads as ineligible.

## How, and why it is safe

`app/utils/storyProgressMigration.ts` reconciles saved marks against the published catalog. It runs on every chapter-catalog load — live fetch, editions cache, and the IndexedDB critical-cache bundle — and after a realtime merge, which unions both sides' objective IDs and can reintroduce a retired one. It is idempotent and acts only on proof:

| Case | Action | Why |
| --- | --- | --- |
| ID is in the published chapter | keep | still valid |
| ID has a proven successor that holds no mark of its own | move | see evidence below |
| ID cannot satisfy the schema's `^[0-9a-f]{24}$` objective shape | drop | can never resolve again, so it would be re-synced forever |
| ID matches the client shape but is not in the chapter | keep | `referenceCoverage.partial` means the published list is a projection |
| Chapter is absent from the catalog | leave the whole chapter alone | absence means an incomplete catalog, not stale progress |

Guards that make a bad catalog harmless:

- Nothing is dropped for a chapter unless **that chapter's own published objective IDs** are all client IDs. A stale overlay, a pre-v1.93 overlay, a failed load, or an empty catalog reduce the pass to a no-op — verified by tests against both real artifacts, not by reading a version string. An earlier revision gated this per *catalog* rather than per *chapter*; that was wrong, because nine of the ten chapters always published client IDs, and the fix is in this branch's history.
- A mark the player already made against the current ID always wins over an inbound legacy mark.
- A realtime merge for an inactive mode is reconciled against the catalog that did load. That stays sound because every decision is either shape-based — no chapter in any mode scope may publish an objective ID of another shape — or validated against an ID that catalog publishes, so a chapter another mode scopes differently can cost a mark but never invent one.

### Alias evidence

`STORY_OBJECTIVE_ID_ALIASES` holds 7 of the 44 retired IDs:

- **Six** whose objective text is byte-identical and occurs exactly once on each side of the change, so no other objective could be meant: `main-3`, `main-4`, `main-7`, `main-9`, `main-14`, `opt-4`.
- **`the-ticket-main-10` → `68e2ecfeb88d405a420774f8`**, which the overlay re-anchored itself in `src/overrides/modes/regular/prestige.json5`, annotated `Was: the-ticket-main-10`. This is the prestige 4/5 gate.

Deliberately unmapped: `main-2` and `main-5` ("Talk to Mr. Kerman" — five candidates in the new data) and `opt-5` ("Talk to Lightkeeper" — two candidates). The other 34 have no counterpart at all. Guessing would record progress the player never made, so those are dropped.

Net effect for a player who had tracked all 44: **7 marks preserved, 37 removed, 0 orphans left behind**, and the prestige gate keeps reading as met.

## Not covered, on purpose

Task IDs get no equivalent pass. The overlay still publishes synthetic task IDs (`new_beginning_prestige_5`, `new_beginning_prestige_6`, plus 36 synthetic objective IDs) as **live** data, so absence from a catalog is not proof of retirement there — unlike story objectives, whose schema forbids the shape outright. When upstream publishes those quests for real, the overlay should declare the replacement (its `Was:` annotation convention already does this in prose) and this module is where that table lands. Promoting the alias table into the overlay itself, so every consumer migrates instead of each inventing its own, is the natural follow-up.

## Verification

Commit `523f2ff6`, clean worktree.

- `pnpm run typecheck` — pass
- `pnpm run lint` — pass (0 errors, 0 warnings)
- `pnpm run lint:fallow --base origin/main` — pass, 0 introduced complexity or dead-code findings
- `pnpm run format:check` — pass
- `pnpm exec vitest run` over the migration, store, realtime, metadata, sync-integration, progress-state and storyline suites — 213 tests, all pass
- Throwaway runs (not committed) against the real artifacts from both overlay revisions: reconciling a blob holding all 44 legacy marks yields exactly 7 migrated / 37 dropped, leaves only client IDs including the prestige gate objective, is a no-op on a second pass, and is a no-op when the loaded catalog is the pre-v1.93 one.

## Summary by CodeRabbit

* **New Features**
  * Added automatic reconciliation of saved story-objective progress across supported game modes.
  * Re-keyed proven retired objective IDs to published replacements when unambiguous.
  * Removed invalid objective IDs while preserving valid, unknown, ambiguous, and unsupported progress.
  * Applied migration after catalog loads and realtime progress updates, including mode-specific catalogs.
  * Reported migrated and dropped objective counts.

* **Documentation**
  * Documented story-progress reconciliation and preservation rules.

* **Tests**
  * Added coverage for migration scenarios, edge cases, idempotence, and catalog-loading behavior.

<details open><summary>Greptile Summary</summary>

The PR adds idempotent reconciliation of saved story-objective IDs with the currently published chapter catalog.
- Migrates seven proven legacy aliases and removes impossible synthetic objective IDs.
- Preserves current marks, unknown client-shaped IDs, and chapters absent from the catalog.
- Runs reconciliation after catalog loads, startup, and realtime merges.
- Adds focused migration, store, and realtime coverage plus system documentation.
</details>

<details open><summary>Confidence Score: 5/5</summary>

The PR appears safe to merge, with no outstanding findings or newly introduced actionable defects.

Both previous findings were manually resolved, and the latest change only clarifies a test name and comment without changing behavior; the current code covers catalog loads and reconciles explicitly targeted modes after realtime merges.
</details>

<details open><summary>Important Files Changed</summary>

| Filename | Overview |
|----------|----------|
| app/utils/storyProgressMigration.ts | Implements guarded, idempotent migration and pruning of saved story-objective IDs. |
| app/stores/useTarkov.ts | Integrates reconciliation into store startup and exposes mode-targeted migration. |
| app/stores/tarkov/realtimeListener.ts | Reconciles the affected mode after realtime union merges before sync resumes. |
| app/stores/useMetadata.ts | Invokes reconciliation whenever a chapter catalog is loaded from live or cached data. |
| app/utils/__tests__/storyProgressMigration.test.ts | Covers aliases, invalid IDs, partial catalogs, conflict handling, and idempotence. |

</details>

<details open><summary>Flowchart</summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Catalog load or realtime merge] --> B[Select target mode progress]
    B --> C{Chapter exists and publishes client IDs?}
    C -- No --> D[Preserve chapter unchanged]
    C -- Yes --> E{Saved objective ID status}
    E -- Published --> F[Keep mark]
    E -- Proven alias and target unmarked --> G[Move mark]
    E -- Valid client-shaped unknown --> F
    E -- Invalid synthetic shape --> H[Drop mark]
    F --> I[Persist reconciled progress]
    G --> I
    H --> I
    D --> I
```
</details>

<sub>Reviews (4): Last reviewed commit: ["test(app): name the mode-argument test f..."](https://github.com/tarkovtracker-org/tarkovtracker/commit/a3a564d73fa4777e95368aa1ff806b22a61fecf4) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=64829494)</sub>

### CODE REVIEW

File: app/stores/useMetadata.ts [hunk 1]
```diff
@@ -256,6 +257,7 @@ const applyCachedEditions = async (
   state.seasonalPerks = perksForMode(seasonalPerks, mode);
   if (!storyChapters.length) return false;
   state.storyChapters = markRaw(sortedStoryChapters(storyChapters));
+  migrateMetadataStoryObjectiveIds();
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply app/stores/useMetadata.ts:260]: Fixed in #850. The IndexedDB critical-cache path also publishes chapters and now invokes the reconciliation, so a signed-out user starting from that bundle is covered alongside the live and editions-cache paths.

File: app/stores/useTarkov.ts [hunk 2]
```diff
@@ -1655,12 +1695,15 @@ export async function initializeTarkovSync() {
     // Repair failed task states for existing users (runs once after data load)
     // This reapplies valid branch failures and clears stale failed flags
     const completionSchemaMigration = tarkovStore.migrateTaskCompletionSchema();
+    const storyObjectiveMigration = tarkovStore.migrateStoryObjectiveIds();
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply app/stores/useTarkov.ts:1719]: Fixed in #850, then narrowed in #855 (`f854369e`). The realtime merge still triggers reconciliation before sync resumes, but only for the mode the loaded catalog belongs to; a merge naming another mode is deferred to that mode's own catalog load.

File: docs/SYSTEMS.md [hunk 3]
```diff
@@ -454,6 +454,19 @@ IDs are upstream-owned: completed marks whose IDs the current chapter no longer
 v1.93 re-keyed all of The Ticket's) are reported for re-checking on the viewer's own progress only,
 never silently remapped or deleted.
 
+Saved story objective marks are reconciled against the published catalog on every chapter-catalog
+load, in all three modes, by `app/utils/storyProgressMigration.ts`. The pass is idempotent and acts
+only on proof: a mark moves when `STORY_OBJECTIVE_ID_ALIASES` records a re-key the published data
+proves (identical unique objective text, or the overlay re-anchoring its own prestige requirement)
+and the target carries no mark of its own; a mark is dropped when its ID cannot satisfy the story
+schema's client-ID shape, so it can never resolve again and would otherwise be re-synced forever; an
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply docs/SYSTEMS.md]: Fixed in #855 (`f854369e`). The cross-mode safety claim is gone from `docs/SYSTEMS.md`. It now states that a catalog is evidence about its own mode only, that another mode is deferred to its own catalog load, and the corrected reconciliation rules including successor-absent preservation.

File: app/stores/useTarkov.ts [hunk 4]
```diff
@@ -351,6 +368,25 @@ const tarkovActions = {
...
+   * Runs on every chapter-catalog load and is idempotent: a proven re-key moves the mark, an ID the
+   * story contract can no longer accept is dropped so it stops being re-saved and re-synced, and an
+   * unrecognized client ID is left alone because the published objective list can be partial.
+   */
+  migrateStoryObjectiveIds(this: TarkovStoreInstance) {
+    const chapters = useMetadataStore().storyChapters;
+    if (!chapters?.length) return { migrated: 0, dropped: 0 };
+    const totals = { migrated: 0, dropped: 0 };
+    for (const mode of GAME_MODE_VALUES) {
+      const result = migrateModeStoryProgress(this[mode], chapters);
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply app/stores/useTarkov.ts]: Fixed in #855 (`f854369e`). An explicitly requested mode is reconciled only when it matches the loaded catalog's mode; otherwise the request is deferred. A store test asserts that a PvP catalog leaves PvE progress untouched and that the same mode reconciles once its own catalog loads.

File: app/stores/useMetadata.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (DysektAI) [reply app/stores/useMetadata.ts:260]: Fixed in #855 (`f854369e`). Both entry points now share the same rule: reconciliation only ever applies the catalog to its own mode, so the realtime path can no longer rewrite another mode's progress.

File: app/utils/storyProgressMigration.ts [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+ * Read from the data rather than from a version string: an overlay that still publishes curated IDs
+ * must not have its saved marks dropped, and a chapter that failed to load must not either.
+ */
+export const publishesClientObjectiveIds = (chapter: StoryChapter): boolean => {
+  const objectiveIds = publishedObjectiveIds(chapter);
+  return objectiveIds.length > 0 && objectiveIds.every((id) => CLIENT_OBJECTIVE_ID.test(id));
+};
+/** A retired ID may only move onto a published ID that holds no mark of its own. */
+const isMovable = (alias: string | undefined, context: MigrationContext): alias is string =>
+  typeof alias === 'string' && context.published.has(alias) && !(alias in context.stored);
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply app/utils/storyProgressMigration.ts:87]: Fixed in #855 (`f854369e`). A retired id whose proven successor the catalog omits is kept rather than dropped, because `referenceCoverage.partial` means the objective list is a projection, and it migrates once the successor appears. Reproduced with the prestige 4/5 Ticket milestone: with `68e2ecfeb88d405a420774f8` withheld, `the-ticket-main-10` survives and then migrates when the catalog publishes it again.

File: app/stores/useTarkov.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (DysektAI) [reply app/stores/useTarkov.ts:1719]: Fixed in #850. Reconciliation runs from the critical-cache hydration path as well as the live and editions-cache loads, and the pass is idempotent, so a late hydration simply reruns it.

File: app/stores/__tests__/useTarkov.storyObjectiveMigration.test.ts [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    const metadataStore = useMetadataStore();
+    metadataStore.storyChapters = [TICKET_CHAPTER];
+    metadataStore.currentGameMode = 'pve';
+    store.pve.storyChapters = storedTicketProgress();
+    expect(store.migrateStoryObjectiveIds()).toEqual({ migrated: 1, dropped: 1 });
+    expect(store.pve.storyChapters['the-ticket']?.objectives).toEqual({
+      [CURRENT_ID]: { complete: true, timestamp: 1000 },
+    });
+  });
+  it('reconciles a mode a realtime merge changed, using the loaded catalog as evidence', () => {
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply app/stores/__tests__/useTarkov.storyObjectiveMigration.test.ts]: Addressed in #855 (`f854369e`). The real-store test now asserts the cross-mode safety directly: a requested mode the loaded catalog does not cover leaves that mode's saved marks untouched, and the same mode reconciles once its own catalog loads. The realtime listener test still asserts the mode argument is passed.
